### ISSUE TITLE
dev: Better covariance learning with `dx.MultivariateNormalTri`

### ISSUE THREAD

**Issue by patel-zeel** [thomaspinder/GPJax#127]
Hi,

I think `tfb.FillScaleTriL` would be a better choice for bijector compared to `tfb.FillTriangular` at https://github.com/thomaspinder/GPJax/blob/ad8b0d55595dc9617e5a92dedb015e587b2f769c/gpjax/variational_families.py#L25 because:
* `tfb.FillTriangular` converts a vector into a triangular matrix without ensuring that the diagonal entries are positive.
* `tfb.FillScaleTriL` uses `tfb.FillTriangular` under the hood and transforms the diagonal with a user-specified bijector e.g. `tfb.Exp` or by default `tfb.Softplus`.

Alternatively, `tfb.CorrelationCholesky` can be used to create a correlation matrix and then transform it to a covariance matrix by an appropriate transformation e.g. Eq. 3.141 in [pml book2](https://probml.github.io/pml-book/book2.html):
<img width="594" alt="image" src="https://user-images.githubusercontent.com/59758528/199453857-3f40e543-d310-4eb4-8e79-427740215912.png">

**Comment by thomaspinder** [comment]
Nice suggestion @patel-zeel. Would you be willing to open a PR for this?

**Comment by patel-zeel** [comment]
Hi @thomaspinder. While testing `tfb.FillScaleTriL`, I noticed it is not working with float64 arrays. I have [raised an issue](https://github.com/tensorflow/probability/issues/1648) in `tfp` for the same.

**Comment by patel-zeel** [comment]
Hi @thomaspinder. After the suggestions given in [the tfp issue](https://github.com/tensorflow/probability/issues/1648), I tried replacing `tfb.FillTriangular` with `tfb.FillScaleTriL`. Currently, I see slower convergence while running the [Sparse Stochastic VI](https://github.com/thomaspinder/GPJax/blob/master/examples/uncollapsed_vi.ipynb) notebook with default `tfb.FillScaleTriL`. When I checked the converged loss after 4000 iterations over multiple random seeds, I found it in the following range:

* `tfb.FillTriangular()`: 500-1000
* `tfb.FillScaleTriL()` with `Softplus` (default) diag_bijector: 6300-6400
* `tfb.FillScaleTriL()` with `Exp` diag_bijector: 5600-5700
* `tfb.FillScaleTriL()` with `Square` diag_bijector: -250-250

I think `Square` is not considered a good bijector because it creates bimodality ([a GPy thread](https://github.com/SheffieldML/GPy/discussions/943#discussioncomment-1405002)). However, in this case, it seems to converge faster than all. I am not sure if there is an applied paper showing the faster convergence (without losing the stability) of using one bijector over the other consistently over multiple models and datasets. I'd love to hear from you if you know of such research.

**Comment by Daniel-Dodd** [comment]
Nice work, @patel-zeel. Thanks for looking into this. I am not aware of such research, but given it seems the most stable, picking `tfb.FillScaleTriL` with `Softplus`  might be a good default option to use in `GPJax`. We could perhaps illustrate the effect of choosing the other bijectors on convergence performance in a new section of the SVGP notebook, to make users aware there are other choices that could give faster convergence, but are less stable. Would love to hear your thoughts on this.

**Comment by patel-zeel** [comment]
Yes, I think stability v/s convergence speed point looks reasonable. I'd go with `tfb.FillScaleTriL` with `Softplus` and create a comparison in a new section of SVGP in a few days. Thank you for your input, @Daniel-Dodd.

**Comment by patel-zeel** [comment]
Hi @Daniel-Dodd, I see that `tensorflow_probability` is now removed from the `variational_families.py` and probably from other places as well. Is it due to complete distraxification or can I use `tfb.FillScaleTriL` to resolve this issue?

**Comment by Daniel-Dodd** [comment]
Hi @patel-zeel, sorry, I forgot to mention in `v0.5`, all the default bijectors moved to `gpjax.config`, so this is why you would see fewer `tensorflow_probability` imports across the package. But, we still support it since it is a dependency of `distrax`, so feel free to replace `tfb.FillTriangular` with `tfb.FillScaleTriL` here to resolve this issue.

**Comment by Daniel-Dodd** [comment]
@patel-zeel, given we build things from a global dictionary `gpjax.config.get_defaults`, could we set the bijector by checking JAX's precision setting?

```python
def get_defaults() -> ConfigDict:
    """Construct and globally register the config file used within GPJax.

    Returns:
        ConfigDict: A `ConfigDict` describing parameter transforms and default values.
    """
    config = ConfigDict(type_safe=False)
    config.key = jr.PRNGKey(123)

    # Covariance matrix stabilising jitter
    config.jitter = 1e-6

    # Find JAX precision (rough way - not sure what the correct way is off the top of my head).
    dtype = jnp.array([0.0]).dtype 

    # Define FillScaleTriL bijector
    FillScaleTriL = tfb.FillScaleTri(diag_shift=jnp.array(config.jitter, dtype=dtype))

   ...
```

**Comment by patel-zeel** [comment]
@Daniel-Dodd, yes, that should have worked. For some reason, it still initializes the float32 version of bijector and fails. I tried the following code after applying your suggestion:

```py
import jax.numpy as jnp
import jax

import tensorflow_probability.substrates.jax as tfp

tfb = tfp.bijectors

import distrax as dx
import gpjax as gpx

# Enable Float64 for more stable matrix inversions.
jax.config.update("jax_enable_x64", True)

eye = jnp.eye(3)
print("eye_dtype", eye.dtype)

FillScaleTriL = dx.Chain(
    [
        tfb.FillScaleTriL(diag_shift=jnp.array(1e-6)),
    ]
)

trial_1 = FillScaleTriL.inverse(eye)  # This works fine

defaults = gpx.config.get_defaults()
trial_2 = defaults.transformations.triangular_transform.inverse(jnp.eye(3))  
# This does not work saying: Tensor conversion 
# requested dtype <class 'numpy.float32'> for array with dtype float64

```

**Comment by Daniel-Dodd** [comment]
@patel-zeel  Oh no. This is really annoying! I would perhaps, suggest coding our own `distrax.bijector` to apply the `Softplus` to the diagonal, then using it in `distrax.Chain` along with `tfb.FillTriangular` (or if this breaks, we can use a distrax implementation of this too)? I've done a quick implementation of transforming the diagonal, though have not moved to the next step of extending this to be compatible with a `distrax.Chain`.

```python
import distrax
import jax
import jax.numpy as jnp

from jaxtyping import Float, Array

class TransformDiagonal(distrax.Bijector):

  def __init__(self, transform: distrax.Bijector = Softplus):
    self.transform = transform


  def forward_and_log_det(self, x: Float[Array, "N N"]):
    idxs = jnp.diag_indices_from(x)
    diag = x[idxs]

    trans_diag, log_det = self.transform.forward_and_log_det(diag)
    
    foward_value = x.at[idxs].set(trans_diag)

    return foward_value, log_det


  def inverse_and_log_det(self, y: Float[Array, "N N"]):
    idxs = jnp.diag_indices_from(y)
    diag = y[idxs]

    trans_diag, log_det = self.transform.inverse_and_log_det(diag)
    
    inverse_value = y.at[idxs].set(trans_diag)

    return inverse_value, log_det


# Example
import jax.random as jr

key = jax.random.PRNGKey(123)
matrix = jr.uniform(key,(3, 3), minval=-1.0, maxval=1.0)
bij = TransformDiagonal()
trans = bij.forward(matrix)
res = bij.inverse(trans)

print(matrix)
print(trans)
print(res)
```

```
[[-0.08364177 -0.4259181  -0.42783093]
 [-0.91206527 -0.8276534  -0.32708693]
 [ 0.68643284  0.67893076 -0.45970702]]
[[ 0.6522005  -0.4259181  -0.42783093]
 [-0.91206527  0.3626089  -0.32708693]
 [ 0.68643284  0.67893076  0.4894806 ]]
[[ 1.0715017  -0.4259181  -0.42783093]
 [-0.91206527  0.89079803 -0.32708693]
 [ 0.68643284  0.67893076  0.9675422 ]]
```

**Comment by patel-zeel** [comment]
Hi @Daniel-Dodd, I am getting back to this after a while because of a deadline :)

I finally caught the culprit, and it is this:
```py
global __config
if not __config:
    __config = config
return __config
```

Since `__config` is created on `import gpjax as gpx` (with `__init__.py`) via https://github.com/JaxGaussianProcesses/GPJax/blob/c50d34d44977e3ef2290576ee1e9b82a69656431/gpjax/kernels.py#L36 it ignores the newly created `__config` after `jax.config.update("jax_enable_x64", True)`. In my opinion, this can be handled in two ways:

* O1: Ensure that `get_defaults()` is never invoked on `import gpjax as gpx`
* O2: If JAX precision changes during the current call of `get_defaults()`, replace the old `__config` with the current config.

O1 might be unnecessary, but O2 can be done easily. What do you think?

Here are a few more notes I'd like to have your views upon:
* N1: I think `if not __config` should be replaced with `if __config is None`. Explicit is better than implicit - The Zen of Python, by Tim Peters :)
* N2: Should the check `if __config is None` be at the top of `get_config` instead of the bottom? For instance, if `__config` is already present in `global` or JAX precision is unchanged, we can ignore the creation of a new config and return old `__config`.
* N3: `JITTER = get_defaults()["jitter"] ` is not used anywhere in `kernels.py` and thus can be removed if desired. Another instance of `get_defaults()` invocation during `import gpjax as gpx` is https://github.com/JaxGaussianProcesses/GPJax/blob/c50d34d44977e3ef2290576ee1e9b82a69656431/gpjax/natural_gradients.py#L35

**Comment by Daniel-Dodd** [comment]
Thanks @patel-zeel, excellent work! 

O1: Agree. This would be good.
O2: Agree. It would be good to ensure the global dictionary is updated if the precision changes (if not already addressed by resolving O1).

N1: Agree. We should ensure we do this throughout the library for other things too. 
N2: Agree.
N3: Agree. The `DEFAULT_JITTER`'s are redundant - we should remove these!

Perhaps, also, it would be good to rename the ``get_defaults()`` convention to ``global_settings()``. This would give greater clarity that the user can override these (and that they are global). Moreover, it might be nice to provide a function to allow the user to revert back to a set of defaults (precision aware) without having to restart their session?


In future, I wish to better harmonise the lower-triangular bijector with the [`LowerTriangularLinearOperator`](https://github.com/JaxGaussianProcesses/JaxLinOp/blob/9d0d6810e3e4493e32337611c81f645b558acb9a/jaxlinop/triangular_linear_operator.py#L26) from the `JaxLinOp` library. If you have any thoughts/ideas on this, I would love to hear them. :)

**Comment by patel-zeel** [comment]
Thanks for the review, @Daniel-Dodd. I'll soon make a PR addressing these issues.
